### PR TITLE
task solved

### PR DESCRIPTION
**Pull request by MykytaKat**

### CODE REVIEW

**Review by OlehPryshchepa — APPROVED** [review]
good job

File: src/main/java/core/basesyntax/service/InvalidDataException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+public class InvalidDataException extends RuntimeException {
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/service/InvalidDataException.java:3]: I'd recommend renaming it to something like 'UserRegistrationException'

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -5,10 +5,32 @@
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int CORRECT_AGE = 18;
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:8]:
```suggestion
    private static final int MIN_AGE = 18;
```

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 3]
```diff
@@ -5,10 +5,32 @@
...
+    private static final int CHARACTER_AMOUNT = 6;
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:9]:
```suggestion
    private static final int MIN_LENGTH = 6;
```

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 4]
```diff
@@ -5,10 +5,32 @@
...
+        if (user.getPassword() == null || user.getPassword().length() < CHARACTER_AMOUNT) {
+            throw new InvalidDataException("Password must contain at least "
+                    + CHARACTER_AMOUNT + " character");
+        }
+        if (user.getId() == null) {
+            throw new InvalidDataException("Id can't be null");
+        }
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new InvalidDataException("Login already taken");
+        }
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:33]: let's move this validation logic to the separate method/methods. Also let's slightly improve exception messages

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 5]
```diff
@@ -5,10 +5,32 @@
...
+        return user;
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:34]: do you really register a user in this method?

File: src/test/java/core/basesyntax/RegistrationServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class RegistrationServiceImplTest {
+
+    private static User user;
+    private static final int DEFAULT_AGE = 25;
+    private static final int INCORRECT_AGE = 15;
+    private static final long DEFAULT_ID = 1L;
```
Comment (sarakhmen) [inline src/test/java/core/basesyntax/RegistrationServiceImplTest.java:18]: let's not validate user's id as it's not required in terms of registration

**Review by sarakhmen — CHANGES_REQUESTED** [review]
Let's improve your solution :) Please also add more tests for different edge cases

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 7]
```diff
@@ -5,10 +5,49 @@
...
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        test(user);
+        return storageDao.add(user);
+    }
+
+    private void test(User user) {
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:20]:
```suggestion
    private void validateUser(User user) {
```

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 8]
```diff
@@ -5,10 +5,49 @@
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MIN_AGE = 18;
+    private static final int MAX_AGE = 100;
+    private static final int MIN_LENGTH = 6;
+    private static final int MAX_LENGTH = 20;
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:11]: as described in common mistakes: 
image

File: src/test/java/core/basesyntax/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.User;
+import core.basesyntax.service.UserRegistrationException;
+import core.basesyntax.service.RegistrationServiceImpl;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class RegistrationServiceImplTest {
+
```
Comment (sarakhmen) [inline src/test/java/core/basesyntax/RegistrationServiceImplTest.java:14]:
```suggestion
```

**Review by sarakhmen — CHANGES_REQUESTED** [review]
Overall good job, but let's slightly improve your solution ;)

File: src/test/java/core/basesyntax/RegistrationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        assertThrows(UserRegistrationException.class,
+                () -> registrationService.register(null),
+                "UserRegistrationException error was expected");
+    }
+
+    @Test
+    void register_nullAge_notOk() {
+        user.setAge(null);
+        assertThrows(UserRegistrationException.class, () -> registrationService.register(user),
+                "Age shouldn't be null");
```
Comment (sarakhmen) [inline src/test/java/core/basesyntax/RegistrationServiceImplTest.java:48]: please fix this 'message' parameter in all places

File: src/test/java/core/basesyntax/RegistrationServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Test
+    void register_ageLessThanEighteen_notOk() {
+        user.setAge(INCORRECT_AGE);
+        assertThrows(UserRegistrationException.class, () -> registrationService.register(user),
+                "Age must be at least 18 years old");
+    }
+
+    @Test
+    void register_PasswordLengthLessThanSix_notOk() {
```
Comment (sarakhmen) [inline src/test/java/core/basesyntax/RegistrationServiceImplTest.java:80]:
```suggestion
    void register_passwordLengthLessThanSix_notOk() {
```
fix the same issue in all places

**Review by sarakhmen — CHANGES_REQUESTED** [review]
Let's improve your solution once again. Also, don't you forget to clear your storage after each test method?

File: src/test/java/core/basesyntax/RegistrationServiceImplTest.java [hunk 12]
```diff
@@ -18,6 +19,7 @@ public class RegistrationServiceImplTest {
     private static final String DEFAULT_LOGIN = "user@gmail.com";
     private static final String DEFAULT_PASSWORD = "abc123";
     private static final String INCORRECT_LOGIN_AND_PASSWORD = "gg";
+    private static final String MASSAGE = "UserRegistrationException error was expected";
```
Comment (maxline) [inline src/test/java/core/basesyntax/RegistrationServiceImplTest.java:22]: fix typo in name please - `MESSAGE`

**Review by maxline — APPROVED** [review]
Well done. Approved with one small remark. Take a look for the future please.
